fuel-vm/src/tests/coins.rs

```diff
@@ -0,0 +1,884 @@
+#[test_case(1, 10, 0 => RunResult::Success((9, 1)); "Can transfer 1 coins to empty")]
+#[test_case(1, 10, 5 => RunResult::Success((9, 6)); "Can transfer 1 coins to non-empty")]
+#[test_case(11, 10, 0 => RunResult::Panic(PanicReason::NotEnoughBalance); "Cannot transfer just over balance coins")]
+#[test_case(Word::MAX, 0, 0 => RunResult::Panic(PanicReason::NotEnoughBalance); "Cannot transfer max over balance coins")]
+#[test_case(1, 1, Word::MAX => RunResult::Panic(PanicReason::BalanceOverflow); "Cannot overflow balance of contract")]
+#[test_case(Word::MAX, Word::MAX, 0 => RunResult::Success((0, Word::MAX)); "Can transfer Word::MAX coins to empty contract")]
+fn transfer_to_contract_external(
+    amount: Word,
+    balance: Word,
+    other_balance: Word,
+) -> RunResult<(Word, Word)> {
+    let contract_id_ptr = 0x11;
+    let asset_id_ptr = 0x12;
+    let reg_amount = 0x13;
+
+    let mut ops = set_full_word(reg_amount.into(), amount);
+
+    ops.extend(&[
+        op::gtf_args(contract_id_ptr, RegId::ZERO, GTFArgs::ScriptData),
+        op::addi(
+            asset_id_ptr,
+            contract_id_ptr,
+            ContractId::LEN.try_into().unwrap(),
+        ),
+        op::tr(contract_id_ptr, reg_amount, asset_id_ptr),
+        op::ret(RegId::ONE),
+    ]);
+
+    let mut test_context = TestBuilder::new(1234u64);
+    let asset_id = AssetId::new([1; 32]);
+
+    let contract = test_context
+        .setup_contract(
+            vec![op::ret(RegId::ONE)],
+            Some((asset_id, other_balance)),
+            None,
+        )
+        .contract_id;
+
+    let script_data: Vec<u8> = contract
+        .to_bytes()
+        .into_iter()
+        .chain(asset_id.to_bytes())
+        .collect();
+
+    let (_, tx, receipts) = test_context
+        .start_script(ops, script_data)
+        .script_gas_limit(1_000_000)
+        .contract_input(contract)
+        .coin_input(asset_id, balance)
+        .fee_input()
+        .contract_output(&contract)
+        .change_output(asset_id)
+        .execute()
+        .into_inner();
+
+    let change = find_change(tx.outputs().to_vec(), asset_id);
+    let result = extract_novalue(&receipts);
+    if !result.is_ok() {
+        assert_eq!(change, balance, "Revert should not change balance")
+    }
+    result.map(|()| {
+        (
+            change,
+            test_context.get_contract_balance(&contract, &asset_id),
+        )
+    })
+}
+
+#[test_case(1, 0, 10, 0 => RunResult::Panic(PanicReason::TransferZeroCoins); "Cannot transfer 0 coins to empty other")]
+#[test_case(1, 0, 10, 5 => RunResult::Panic(PanicReason::TransferZeroCoins); "Cannot transfer 0 coins to non-empty other")]
+#[test_case(0, 0, 10, 0 => RunResult::Panic(PanicReason::TransferZeroCoins); "Cannot transfer 0 coins to self")]
+#[test_case(1, 1, 10, 0 => RunResult::Success((9, 1)); "Can transfer 1 coins to other")]
+#[test_case(0, 1, 10, 0 => RunResult::Success((10, 0)); "Can transfer 1 coins to self")]
```

Comment:
It seems the self transfers doesn't affect values in `Success`. Maybe we want to list all balances of all contract there for clarity of the expected result.